### ISSUE TITLE
recursive conditional type fails to narrow when using `...` operator

### ISSUE THREAD

**Issue by DetachHead** [microsoft/TypeScript#46176]
# Bug Report

### 🔎 Search Terms

recursive conditional type narrow array 

### 🕗 Version & Regression Information

4.1.5 - 4.5.0-dev.20211002

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.4.3#code/C4TwDgpgBA+gKgVzAGwgeQGYB4BQV9RzgQA0eBAMhAHYDmwAFlBAB7A0AmAzlNQgLYAjCACcyBKAGEEIkTWABBWQEMQzNpx5FIAbQC6OAHxQAvFJlzqilSB0ByVHUZ2969tW5QqThuXwB+c1l5JRFVPygALlhEFHRsbVIvGnoGEigdRPSAOlzpYKtQ1T1DHBxWMAB7EWAoUEhCJFRMLCzknzdNXgFhEWMzPiFRTo8eb1TAxP1o+Cb41uJ08cZ0-VKceugFDg4sADkBAEYRz0He9IP+ACYTnjPRfoyIgHpnqABLDCgQSoQoLgA7u9gABjJjASpQAAs2QArN9fg5kFBaBBaspqMxZNUoAxRNAOAhoBDccBgGAuJFXrRgQwEIJsiDKvxnvx3iCRJUuJUMMBnokAMoc95gPnvLhcIlcZ5QgBshwA7IcIrlsrFmthqBAAG6iC5HQzpFW5dXzLW6sRQS5XQx6eyOVIuMqbKAKCWiYCXXqtW7dIZ9UyEMqvUSckTRRJQOzbXZwdI2uxQDiVCB3Sq1LjKYDijBqRjQJnULjAMLvKxR+4iOzZAAUVwAzFCoQBKDbEKAAMUqlR9rHcpx6D0Dbq4Hq9oiwMda8cMpSAA)

### 💻 Code

```ts
type _TupleOf<
    Type,
    Length extends number,
    CurrentArray extends Type[]
> = CurrentArray['length'] extends Length
    ? CurrentArray
    : _TupleOf<Type, Length, [Type, ...CurrentArray]>

export type TupleOf<Type, Length extends number> = number extends Length? Type[]: _TupleOf<Type, Length, []>

type Add<Num1 extends number, Num2 extends number> = [
    ...TupleOf<never, Num1>, 
    ...TupleOf<never, Num2>]['length']

type AssertNumber<T extends number> = T

//error: Type 'Add<T, 2>' does not satisfy the constraint 'number'.(2344)
type Foo<T extends number> = AssertNumber<Add<T, 2>>

```

### 🙁 Actual behavior
```
Type 'Add<T, 2>' does not satisfy the constraint 'number'.(2344)
```

### 🙂 Expected behavior

no error

**Comment by ruohola** [comment]
> **🕗 Version & Regression Information**
>
> 4.1.5 to 4.5.0-dev.20211002

This doesn't seem to have worked in 4.1.5, or actually ever?

**Comment by DetachHead** [comment]
Yeah sorry that was a bit lazy of me, that was just the earliest version on playground I tested against before other errors started to show up

**Comment by DetachHead** [comment]
potentially the same issue, not sure:
```ts
type Foo<U> = U extends [infer I]? I : never;

type AssertString<T extends string> = T

// Type 'Foo<T>' does not satisfy the constraint 'string'.
//  Type 'unknown' is not assignable to type 'string'.(2344)
type Bar<T extends string> = AssertString<Foo<T>>
type Baz = AssertString<Foo<string>>
```
https://www.typescriptlang.org/play?#code/C4TwDgpgBAYg9nAPAVQHxQLxWVCAPYCAOwBMBnKAbQEsiAzCAJygEkBdAflagC4oiIANyYBuAFBjQkKAEEyZJsADKwRrQDmiACq4CxclDKqN6LFokB6C1C3hoAcnhItqe1BJwIFInGCGAhsDUZHQgUMAAFtAAxnBERoz+tH72CRr2AHRiVlA2dlD2AK5EANY+AO5EbsH8vlD+8tTqRP4ARgA20MBw4fmpxkTqmQAUAEwAzAAskwCUkvkAQv6M2rqEpBRpg6ay8ooqaoOITtqoqPPSSwBemLsKjMoDmidb6mdAA

**Comment by andrewbranch** [comment]
Duplicate of https://github.com/microsoft/TypeScript/issues/44433 and many other generic conditional type / indexed access type issues—conditional types and indexed access types do not resolve according to their constraints and are not generally relatable when generic.
